File: week-06/scripts/07_parallel_sort_cpu.py
#!/usr/bin/env python3
"""Measure a real sequential-versus-multiprocess CPU sorting path.

This is the no-compiler fallback for the Week 6 parallelism lesson.  It uses
separate Python processes (not threads) so chunk sorting can really run on
multiple CPU logical processors.  The timed parallel operation includes
dispatching chunks and the final merge: that is intentionally the cost a
caller pays, rather than a flattering but incomplete chunk-sort number.
"""

from __future__ import annotations

import argparse
import csv
import heapq
import math
import multiprocessing as mp
import random
import statistics
import time
from pathlib import Path


SIZES = (32, 256, 2_048, 16_384, 131_072, 1_048_576, 2_097_152)


def sort_chunk(chunk: list[int]) -> list[int]:
    """Worker entry point.  Kept module-level so the pool can pickle it."""
    chunk.sort()
    return chunk


def source_for(n: int) -> list[int]:
    """Deterministic source values: every backend receives the same values."""
    rng = random.Random(0xC0FFEE + n)
    return [rng.getrandbits(32) for _ in range(n)]


def repeats_for(n: int, base: int) -> int:
    """Use enough samples for tiny values without making large runs excessive."""
    if n <= 2_048:
        return max(base, 31)
    if n <= 16_384:
        return max(base, 11)
    if n <= 131_072:
        return max(base, 5)
    return base


def batch_for(n: int) -> int:
    """Keep a tiny-operation sample above timer noise, for both backends."""
    if n <= 32:
        return 3
    if n <= 256:
        return 3
    return 1


def chunks_for(source: list[int], workers: int) -> list[list[int]]:
    chunk_size = math.ceil(len(source) / workers)
    return [source[start : start + chunk_size] for start in range(0, len(source), chunk_size)]


def median_ms(samples: list[float]) -> float:
    return statistics.median(samples) * 1_000


def benchmark_sequential(source: list[int], expected: list[int], repeats: int, batch: int) -> float:
    samples: list[float] = []
    for _ in range(repeats):
        work_items = [list(source) for _ in range(batch)]  # Input preparation is outside timing.
        start = time.perf_counter()
        for work in work_items:
            work.sort()
        samples.append((time.perf_counter() - start) / batch)
        if any(work != expected for work in work_items):
            raise RuntimeError("sequential sort failed its exact-output check")
    return median_ms(samples)


def benchmark_parallel(
    source: list[int], expected: list[int], pool: mp.pool.Pool, workers: int, repeats: int, batch: int
) -> float:
    samples: list[float] = []
    for _ in range(repeats):
        # Just like the sequential list copy, forming source chunks is outside the
        # measured sort.  Sending them to workers, coordinating them, and merging
        # their answers are inside: those are real parallel-sort overheads.
        chunk_batches = [chunks_for(source, workers) for _ in range(batch)]
        start = time.perf_counter()
        results = []
        for chunks in chunk_batches:
            sorted_chunks = pool.map(sort_chunk, chunks)
            results.append(list(heapq.merge(*sorted_chunks)))
        samples.append((time.perf_counter() - start) / batch)
        if any(result != expected for result in results):
            raise RuntimeError("parallel sort failed its exact-output/multiset check")
    return median_ms(samples)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output", required=True, type=Path)
    parser.add_argument("--threads", required=True, type=int)
    parser.add_argument("--repeats", default=3, type=int)
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    workers = max(2, args.threads)
    base_repeats = max(1, args.repeats)
    rows: list[dict[str, object]] = []

    # Fork is available on this Linux/WSL teaching host and avoids a first-use
    # interpreter import spike.  The pool remains warm while every measured map
    # and merge still includes the parallel operation's coordination cost.
    context = mp.get_context("fork")
    with context.Pool(processes=workers) as pool:
        warm = source_for(16_384)
        warm_expected = sorted(warm)
        benchmark_sequential(warm, warm_expected, 1, 1)
        benchmark_parallel(warm, warm_expected, pool, workers, 1, 1)

        for n in SIZES:
            repeats = repeats_for(n, base_repeats)
            batch = batch_for(n)
            print(f"n={n:,} ({repeats} median samples × {batch} operations/sample) ...", flush=True)
            source = source_for(n)
            expected = sorted(source)  # Oracle is outside every timed sample.
            sequential = benchmark_sequential(source, expected, repeats, batch)
            parallel = benchmark_parallel(source, expected, pool, workers, repeats, batch)
            rows.extend(
                (
                    {"backend": "cpu_sequential", "n": n, "workers": 1, "median_ms": f"{sequential:.6f}"},
                    {"backend": "cpu_parallel", "n": n, "workers": workers, "median_ms": f"{parallel:.6f}"},
                )
            )
            print(
                f"  sequential {sequential:.3f} ms | parallel {parallel:.3f} ms "
                f"| speedup {sequential / parallel:.3f}x",
                flush=True,
            )

    args.output.parent.mkdir(parents=True, exist_ok=True)
    temporary = args.output.with_suffix(args.output.suffix + ".tmp")
    with temporary.open("w", newline="") as file:
        writer = csv.DictWriter(
            file,
            fieldnames=("backend", "n", "workers", "median_ms"),
            lineterminator="\n",
        )
        writer.writeheader()
        writer.writerows(rows)
    temporary.replace(args.output)
    print(f"Wrote {args.output}")


if __name__ == "__main__":
    main()

File: week-06/scripts/07_parallel_sort_cpu.cpp
// Week 6: CPU sequential vs parallel sort benchmark.
//
// Sequential path: std::sort
// Parallel path: GNU libstdc++ parallel multiway mergesort using OpenMP.
//
// Compile:
//   g++ -O3 -std=c++17 -fopenmp 07_parallel_sort_cpu.cpp -o parallel_sort_cpu
//
// This deliberately times sorting only. Input generation and vector copying happen
// outside the timer so we can discuss algorithm/runtime overhead separately.

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <fstream>
#include <functional>
#include <iomanip>
#include <iostream>
#include <parallel/algorithm>
#include <parallel/tags.h>
#include <random>
#include <string>
#include <thread>
#include <vector>
#include <omp.h>

using Clock = std::chrono::steady_clock;

static std::vector<std::size_t> sizes() {
    return {10, 100, 1000, 10000, 100000, 1000000, 5000000, 10000000};
}

static std::vector<std::uint32_t> make_data(std::size_t n, std::uint32_t seed) {
    std::mt19937 rng(seed);
    std::uniform_int_distribution<std::uint32_t> dist;
    std::vector<std::uint32_t> values(n);
    for (auto &v : values) v = dist(rng);
    return values;
}

static int repeats_for(std::size_t n, int base) {
    if (n <= 1000) return std::max(base, 51);
    if (n <= 10000) return std::max(base, 11);
    return base;
}

template <typename F>
static double median_ms(
    const std::vector<std::uint32_t>& source,
    int repeats,
    F sorter
) {
    std::vector<double> times;
    times.reserve(repeats);

    for (int r = 0; r < repeats; ++r) {
        auto work = source;  // outside timer

        const auto t0 = Clock::now();
        sorter(work);
        const auto t1 = Clock::now();

        if (!std::is_sorted(work.begin(), work.end())) {
            throw std::runtime_error("sort produced incorrect output");
        }

        times.push_back(
            std::chrono::duration<double, std::milli>(t1 - t0).count()
        );
    }

    std::sort(times.begin(), times.end());
    return times[times.size() / 2];
}

int main(int argc, char** argv) {
    std::string output = "parallel_sort_results.csv";
    int threads = std::min(16, std::max(2, omp_get_max_threads()));
    int base_repeats = 3;

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];

        if (arg == "--output" && i + 1 < argc) {
            output = argv[++i];
        } else if (arg == "--threads" && i + 1 < argc) {
            threads = std::max(2, std::stoi(argv[++i]));
        } else if (arg == "--repeats" && i + 1 < argc) {
            base_repeats = std::max(1, std::stoi(argv[++i]));
        } else {
            std::cerr
                << "Usage: " << argv[0]
                << " [--output FILE] [--threads N] [--repeats N]\n";
            return 2;
        }
    }

    std::ofstream out(output);
    if (!out) {
        std::cerr << "Could not open " << output << " for writing.\n";
        return 1;
    }

    out << "backend,n,workers,median_ms\n";
    out << std::fixed << std::setprecision(6);

    std::cerr << "CPU hardware threads reported: "
              << std::thread::hardware_concurrency() << "\n";
    std::cerr << "OpenMP max threads: " << omp_get_max_threads() << "\n";
    std::cerr << "Parallel workers requested: " << threads << "\n";

    // Warm both code paths before collecting measurements.
    auto warm = make_data(20000, 7);
    auto warm_seq = warm;
    std::sort(warm_seq.begin(), warm_seq.end());

    auto warm_par = warm;
    __gnu_parallel::sort(
        warm_par.begin(),
        warm_par.end(),
        std::less<std::uint32_t>(),
        __gnu_parallel::multiway_mergesort_tag(threads)
    );

    for (const auto n : sizes()) {
        const int repeats = repeats_for(n, base_repeats);
        std::cerr << "n=" << n << " (" << repeats << " repeats) ...\n";

        const auto source =
            make_data(n, static_cast<std::uint32_t>(0xC0FFEEu + n));

        const double sequential = median_ms(
            source,
            repeats,
            [](auto& work) {
                std::sort(work.begin(), work.end());
            }
        );

        const double parallel = median_ms(
            source,
            repeats,
            [threads](auto& work) {
                __gnu_parallel::sort(
                    work.begin(),
                    work.end(),
                    std::less<std::uint32_t>(),
                    __gnu_parallel::multiway_mergesort_tag(threads)
                );
            }
        );

        out << "cpu_sequential," << n << ",1," << sequential << "\n";
        out << "cpu_parallel," << n << "," << threads << "," << parallel << "\n";
        out.flush();

        std::cerr
            << "  sequential " << sequential << " ms"
            << " | parallel " << parallel << " ms"
            << " | speedup " << (sequential / parallel) << "x\n";
    }

    std::cerr << "Wrote " << output << "\n";
    return 0;
}
